### PR TITLE
storage: only acquire read lock when checking for in-progress merge

### PR DESCRIPTION
**Pull request by benesch**
Replica.mu is a reader-writer lock, so we can be slightly more efficient
when we check to see whether a merge is in progress by only acquiring a
read lock.

Release note: None

### CODE REVIEW

**Review by tbg — APPROVED** [review]
:lgtm:

> Reviewed 1 of 1 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27400#-:-LH8wSqY7vE5ixsob9Xa:bnfp4nl)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

**Review by nvb — APPROVED** [review]

**Comment by benesch** [comment]
Updated to also fix the race that's failing https://github.com/cockroachdb/cockroach/issues/27403 and friends. PTAQL!

File: pkg/storage/client_merge_test.go [hunk 1]
```diff
@@ -186,25 +187,25 @@ func TestStoreRangeMergeWithData(t *testing.T) {
...
 	sc.TestingKnobs.DisableReplicateQueue = true
 
 	// Maybe inject some retryable errors when the merge transaction commits.
 	sc.TestingKnobs.TestingRequestFilter = func(ba roachpb.BatchRequest) *roachpb.Error {
-		if retries > 0 {
+		if atomic.LoadInt64(&retries) > 0 {
 			for _, req := range ba.Requests {
 				if et := req.GetEndTransaction(); et != nil && et.InternalCommitTrigger.GetMergeTrigger() != nil {
-					retries--
+					atomic.AddInt64(&retries, -1)
```
Comment (nvanbenschoten) [inline pkg/storage/client_merge_test.go]: Do we want some kind of CAS approach here to avoid letting `retries` drop below 0? Does that matter?

**Review by nvb — APPROVED** [review]

**Review by tbg — APPROVED** [review]

**Review by bdarnell — APPROVED** [review]
> Reviewed 1 of 1 files at r1, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27400)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

**Comment by benesch** [comment]
bors r=bdarnell,tschottdorf,nvanbenschoten

**Pull request merged**